okhttp/src/main/java/okhttp3/EventListener.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (C) 2017 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3;
+
+import java.net.InetAddress;
+import java.util.List;
+
+/**
+ * Observes, modifies, and potentially short-circuits requests going out and the corresponding
+ * responses coming back in. Typically interceptors add, remove, or transform headers on the request
+ * or response.
+ */
+public class EventListener {
```

Comment:
Very insightful request, I am still wrapping my head around this a little.  These methods would almost have to be re-entrant, no?  If OkHttp had to serialize these across workers it would probably have a significant performance impact.

Ideally if a *Start is called I think it is reasonable that *End would be called.  For the most part I could see that working but presently I have a tough time imagining an implementation that guarantees a `fetchEnd` for every `fetchStart`